contracts/utils/RolesBase.sol

```diff
@@ -0,0 +1,282 @@
+ */
+contract RolesBase is IRolesBase {
+    bytes32 internal constant ROLES_PREFIX = keccak256('roles');
+    bytes32 internal constant PROPOSE_ROLES_PREFIX = keccak256('propose-roles');
+
+    /**
+     * @notice Modifier that throws an error if called by any account missing the role.
+     */
+    modifier onlyRole(uint8 role) {
+        if (!_hasRole(_getRoles(msg.sender), role)) {
+            revert MissingRole(msg.sender, role);
+        }
+
+        _;
+    }
+
+    /**
+     * @notice Modifier that throws an error if called by an account without all the roles.
+     */
+    modifier withEveryRole(uint8[] memory roles) {
+        if (!_hasAllTheRoles(_getRoles(msg.sender), roles)) {
+            revert MissingAllRoles(msg.sender, roles);
+        }
+
+        _;
+    }
+
+    /**
+     * @notice Modifier that throws an error if called by an account without any of the roles.
+     */
+    modifier withAnyRole(uint8[] memory roles) {
+        if (!_hasAnyOfRoles(_getRoles(msg.sender), roles)) {
+            revert MissingAnyOfRoles(msg.sender, roles);
+        }
+
+        _;
+    }
+
+    /**
+     * @notice Checks if an account has a role.
+     * @param account The address to check
+     * @param role The role to check
+     * @return True if the account has the role, false otherwise
+     */
+    function hasRole(address account, uint8 role) public view returns (bool) {
+        return _hasRole(_getRoles(account), role);
+    }
+
+    /**
+     * @notice Internal function to convert an array of roles to a uint256.
+     * @param roles The roles to convert
+     * @return accountRoles The roles in uint256 format
+     */
+    function _toAccountRoles(uint8[] memory roles) internal pure returns (uint256) {
+        uint256 length = roles.length;
+        uint256 accountRoles;
+
+        for (uint256 i = 0; i < length; ++i) {
+            accountRoles |= (1 << roles[i]);
+        }
+
+        return accountRoles;
+    }
+
+    /**
+     * @notice Internal function to get the key of the roles mapping.
+     * @param account The address to get the key for
+     * @return key The key of the roles mapping
+     */
+    function _rolesKey(address account) internal view virtual returns (bytes32 key) {
+        return keccak256(abi.encodePacked(ROLES_PREFIX, account));
+    }
+
+    /**
+     * @notice Internal function to get the roles of an account.
+     * @param account The address to get the roles for
+     * @return accountRoles The roles of the account in uint256 format
+     */
+    function _getRoles(address account) internal view returns (uint256 accountRoles) {
+        bytes32 key = _rolesKey(account);
+        assembly {
+            accountRoles := sload(key)
+        }
+    }
+
+    /**
+     * @notice Internal function to set the roles of an account.
+     * @param account The address to set the roles for
+     * @param accountRoles The roles to set
+     */
+    function _setRoles(address account, uint256 accountRoles) private {
```

Comment:
Why is this private?

Reply:
The idea here is not to write to the storage directly bypassing event emission.